### PR TITLE
Move optimizations to a new module, and add Section and annotations

### PR DESCRIPTION
**Pull request by IgorBaratta**
- Introduce `Lnode.Section`:
  - A section is a collection of statements that form a coherent segment of the kernel, such as jacobian computation or coefficient evaluation at quadrature points.
  - A section has additional details about input/output and includes annotations for potential optimization strategies.

- Optimization strategies are now decoupled from the initial code generation (LNodes) and are instead applied as a subsequent step. This approach utilizes the annotations provided within each section to tailor optimizations.

- Planned optimization to include:
  - [x] Loop Fusion: Merging loops across sections sharing the same name to improve cache performance and reduce overhead.
  - [x] Loop-Invariant Code Motion: Optimizing rank 2 forms by moving calculations that are constant within a loop outside of it, reducing redundant computations.
  - [ ] Sum Factorization: Applying this technique to jacobian, coefficient, and tensor assembly to reduce the computational complexity.

### CODE REVIEW

File: ffcx/codegeneration/lnodes.py [hunk 1]
```diff
@@ -720,31 +722,55 @@ def as_statement(node):
...
+
+    def __init__(self, name: str, statements: List, input: Optional[List] = None,
+                 output: Optional[List] = None, annotations: Optional[List] = None):
+        self.name = name
+        self.statements = [as_statement(st) for st in statements]
+        self.annotations = annotations or []
+        self.input = input or []
+        self.output = output or []
+
+        # Convert all inputs to strings
```
Comment (chrisrichardson) [inline ffcx/codegeneration/lnodes.py:744]: Can we keep `input` and `ouput` as `Symbols` - so we can see the type, and maybe dimension?

File: ffcx/ir/representation.py [hunk 2]
```diff
@@ -629,6 +629,10 @@ def _compute_expression_ir(expression, index, prefix, analysis, options, visuali
     # Extract dimensions for elements of arguments only
     arguments = ufl.algorithms.extract_arguments(expression)
     argument_elements = tuple(f.ufl_function_space().ufl_element() for f in arguments)
+
```
Comment (chrisrichardson) [inline ffcx/ir/representation.py:632]: remove this debug

File: ffcx/codegeneration/optimizer.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+        List of LNodes to fuse.
+
+    Returns
+    -------
+    list of LNodes
+        Fused list of LNodes.
+
+    """
+    loops = defaultdict(list)
+    pre_loop = []
```
Comment (chrisrichardson) [inline ffcx/codegeneration/optimizer.py:96]: Better to just call this `code` not `pre_loop`?

File: ffcx/codegeneration/C/c_implementation.py [hunk 4]
```diff
@@ -178,6 +178,16 @@ def _build_initializer_lists(self, values):
     def format_statement_list(self, slist) -> str:
         return "".join(self.c_format(s) for s in slist.statements)
 
+    def format_section(self, section) -> str:
```
Comment (jorgensd) [inline ffcx/codegeneration/C/c_implementation.py:181]: Add type hint on section for easy lookup of content.

File: ffcx/codegeneration/optimizer.py [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+    output_code = []
+    for statement in code.statements:
+        if isinstance(statement, L.ForRange):
+            id = (statement.index, statement.begin, statement.end)
+            loops[id].append(statement.body)
+        else:
+            output_code.append(statement)
+
+    for info, body in loops.items():
+        index, begin, end = info
```
Comment (chrisrichardson) [inline ffcx/codegeneration/optimizer.py:107]:
```suggestion
        index, begin, end = info
```

File: ffcx/codegeneration/optimizer.py [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+        loop = L.ForRange(index, begin, end, body)
+        output_code.append(loop)
```
Comment (chrisrichardson) [inline ffcx/codegeneration/optimizer.py:109]:
```suggestion
    for range, body in loops.items():
        output_code.append(L.ForRange(*range, body))
```

File: ffcx/codegeneration/access.py [hunk 7]
```diff
@@ -350,3 +362,48 @@ def facet_edge_vectors(self, e, mt, tabledata, num_points):
...
+
+        if tabledata.is_piecewise:
+            iq_global_index = L.LiteralInt(0)
+
+        # FIXME: Hopefully tabledata is not permuted when applying sum
+        # factorization
+        if tabledata.is_permuted:
+            qp = self.symbols.quadrature_permutation[0]
+            if restriction == "-":
+                qp = self.symbols.quadrature_permutation[1]
```
Comment (jorgensd) [inline ffcx/codegeneration/access.py:395]: We should really either:
1. Use enums for restrictions in FFCx
2. Use enums for restrictions in UFL (which could have a string representation: https://docs.python.org/3/library/enum.html#enum.StrEnum).
@mscroggs what do you think?

File: ffcx/codegeneration/definitions.py [hunk 8]
```diff
@@ -61,51 +49,53 @@ def create_dof_index(tabledata, dof_index_symbol):
...
+                               ufl.geometry.ReferenceCellVolume: self.pass_through,
+                               ufl.geometry.ReferenceFacetVolume: self.pass_through,
+                               ufl.geometry.ReferenceCellEdgeVectors: self.pass_through,
+                               ufl.geometry.ReferenceFacetEdgeVectors: self.pass_through,
+                               ufl.geometry.ReferenceNormal: self.pass_through,
+                               ufl.geometry.CellOrientation: self.pass_through,
+                               ufl.geometry.FacetOrientation: self.pass_through}
+
+    def get(self, mt, tabledata, quadrature_rule, access):
+        # Call appropriate handler, depending on the type of terminal
```
Comment (jorgensd) [inline ffcx/codegeneration/definitions.py:80]: Type hint + make this a docstring?

File: ffcx/codegeneration/definitions.py [hunk 9]
```diff
@@ -227,23 +210,12 @@ def spatial_coordinate(self, e, mt, tabledata, quadrature_rule, access):
...
-        # TODO: Generate more efficient inline code for Max/MinCell/FacetEdgeLength
-        #       and CellDiameter here rather than lowering these quantities?
-        return [], []
-
-    def _pass(self, *args, **kwargs):
-        """Return nothing."""
-        return [], []
+        return self._define_coordinate_dofs_lincomb(mt, tabledata, quadrature_rule, access)
+
+    def pass_through(self, mt, tabledata, quadrature_rule, access):
```
Comment (jorgensd) [inline ffcx/codegeneration/definitions.py:219]:
```suggestion
    def pass_through(self, mt, tabledata, quadrature_rule, access) -> list
```
Maybe?

File: ffcx/codegeneration/integral_generator.py [hunk 10]
```diff
@@ -235,85 +231,78 @@ def declare_table(self, name, table):
...
+        tensor_comp, intermediates_fw = self.generate_dofblock_partition(quadrature_rule)
+
+        # Check if we only have Section objects
+        inputs = []
+        for definition in definitions:
+            assert isinstance(definition, L.Section)
+            inputs += definition.output
+
+        for tc in tensor_comp:
+            assert isinstance(tc, L.Section)
```
Comment (jorgensd) [inline ffcx/codegeneration/integral_generator.py:247]:
```python
assert [isinstance(tc, L.Section) for tc in tensor_comp].all()
```
+ move up to 239

File: ffcx/codegeneration/integral_generator.py [hunk 11]
```diff
@@ -235,85 +231,78 @@ def declare_table(self, name, table):
...
-            # cache
+            # Generate code only if the expression is not already in cache
             if not self.get_var(quadrature_rule, v):
                 if v._ufl_is_literal_:
                     vaccess = L.ufl_to_lnodes(v)
-                elif mt is not None:
-                    # All finite element based terminals have table
-                    # data, as well as some, but not all, of the
-                    # symbolic geometric terminals
+                elif mt:
```
Comment (jorgensd) [inline ffcx/codegeneration/integral_generator.py:295]: Replace with assignment expression: https://docs.python.org/3/whatsnew/3.8.html#assignment-expressions

File: ffcx/codegeneration/integral_generator.py [hunk 12]
```diff
@@ -235,85 +231,78 @@ def declare_table(self, name, table):
...
-                        access = predef[0].symbol.name
-                        pre_definitions[str(access)] = predef
-
-                    # Store definitions of terminals in list
-                    assert isinstance(vdef, list)
-                    definitions[str(vaccess)] = vdef
+                    vaccess = self.backend.access.get(mt, tabledata, quadrature_rule)
+                    vdef = self.backend.definitions.get(mt, tabledata, quadrature_rule, vaccess)
+
+                    if isinstance(vdef, L.Section):
```
Comment (jorgensd) [inline ffcx/codegeneration/integral_generator.py:303]: Change to assert as it should always be true (according to the almighty Baratta)

File: ffcx/codegeneration/integral_generator.py [hunk 13]
```diff
@@ -350,30 +339,16 @@ def generate_partition(self, symbol, F, mode, quadrature_rule):
                     else:
                         # Record assignment of vexpr to intermediate variable
                         j = len(intermediates)
-                        if use_symbol_array:
-                            vaccess = symbol[j]
-                            intermediates.append(L.Assign(vaccess, vexpr))
-                        else:
-                            vaccess = L.Symbol("%s_%d" % (symbol.name, j))
-                            intermediates.append(L.VariableDecl(vaccess, vexpr))
+                        vaccess = L.Symbol("%s_%d" % (symbol.name, j), dtype=L.DataType.SCALAR)
```
Comment (jorgensd) [inline ffcx/codegeneration/integral_generator.py:342]: fstring?

File: ffcx/codegeneration/lnodes.py [hunk 14]
```diff
@@ -722,31 +724,68 @@ def as_statement(node):
...
 
+class Annotation(Enum):
+    fuse = 1        # fuse loops in section
+    unroll = 2      # unroll loop in section
+    licm = 3        # loop invariant code motion
+    factorize = 4   # apply sum factorization
+
+
+def is_declaration(node):
+    return isinstance(node, VariableDecl) or isinstance(node, ArrayDecl)
```
Comment (jorgensd) [inline ffcx/codegeneration/lnodes.py:743]:
```suggestion
def is_declaration(node) -> bool:
    return isinstance(node, VariableDecl) or isinstance(node, ArrayDecl)
```

File: ffcx/codegeneration/lnodes.py [hunk 15]
```diff
@@ -846,11 +885,17 @@ def is_simple_inner_loop(code):
     return False
 
 
+def depth(code):
```
Comment (jorgensd) [inline ffcx/codegeneration/lnodes.py:888]:
```suggestion
def depth(code) -> int:
```

File: ffcx/codegeneration/access.py (hunk 7 above)
Comment (IgorBaratta) [reply ffcx/codegeneration/access.py:395]: Should this be discussed in the UFL issue tracker?

File: ffcx/codegeneration/access.py [hunk 16]
```diff
@@ -350,3 +362,48 @@ def facet_edge_vectors(self, e, mt, tabledata, num_points):
...
         return 1
+
+    def table_access(self, tabledata: UniqueTableReferenceT, entitytype: str, restriction: str,
+                     quadrature_index: L.MultiIndex, dof_index: L.MultiIndex):
+        """
+        Access element table for given entity, quadrature point, and dof index.
+
+        Args:
+            tabledata: Table data object
+            entitytype: Entity type ("cell", "facet", "vertex")
```
[the remark this answers is not in the document]
Comment (IgorBaratta) [reply ffcx/codegeneration/access.py:373]: I agree, let's open an issue for that.
This PR doesn't touch the IR, and this where the entity type is defined.

File: ffcx/codegeneration/access.py (hunk 7 above)
Comment (jorgensd) [reply ffcx/codegeneration/access.py:395]: Yes, do you mind adding an issue?

**Review by jorgensd — APPROVED** [review]

File: ffcx/codegeneration/definitions.py [hunk 17]
```diff
@@ -114,6 +107,11 @@ def coefficient(self, t, mt, tabledata, quadrature_rule, access):
         iq_symbol = self.symbols.quadrature_loop_index
         ic_symbol = self.symbols.coefficient_dof_sum_index
 
+        if isinstance(access, str):
```
Comment (chrisrichardson) [inline ffcx/codegeneration/definitions.py:110]: Contradicts line 101: `access` is always `str` ?

Comment (IgorBaratta) [reply ffcx/codegeneration/definitions.py:110]: Good point. It's not uniform, from expression generator it's a symbol, and from integral generator it's a string.
I'll make it consistent.

**Review by chrisrichardson — APPROVED** [review]

**Pull request merged**